Add tests for `NoDataFlowLoops` rule

The `NoDataFlowLoops` semantic validation rule was added to the codebase without tests. Let's add them, so it's easier to verify that it is working correctly and we improve the quality of the project.

https://github.com/Gelio/CAL-web/blob/main/backend/sirius-web-services/src/main/java/org/eclipse/sirius/web/services/validation/semantic/rules/NoDataFlowLoops.java

After adding a single test I have noticed that I get error messages about code style issues when running `./mvnw test`:

```
[ERROR] Tests run: 6, Failures: 2, Errors: 1, Skipped: 0, Time elapsed: 0.061 s <<< FAILURE! - in org.eclipse.sirius.web.services.architecture.ImmutableTests
[ERROR] immutableClassesShouldHaveAPrivateConstructor  Time elapsed: 0.022 s  <<< FAILURE!
java.lang.AssertionError:
Architecture Violation [Priority: MEDIUM] - Rule 'classes that reside in a package 'org.eclipse.sirius.web.services..' and are annotated with @Immutable and are not annotated with @GraphQLInputObjectType should have only private constructors' was violated (2 times):
Constructor <org.eclipse.sirius.web.services.api.projects.UnzippedProject.<init>()> is not private in (UnzippedProject.java:31)
Constructor <org.eclipse.sirius.web.services.api.representations.RepresentationDescriptor.<init>()> is not private in (RepresentationDescriptor.java:27)

[ERROR] immutableClassesShouldBeFinal  Time elapsed: 0.005 s  <<< FAILURE!
java.lang.AssertionError:
Architecture Violation [Priority: MEDIUM] - Rule 'classes that reside in a package 'org.eclipse.sirius.web.services..' and are annotated with @Immutable should have modifier FINAL' was violated (2 times):
Class <org.eclipse.sirius.web.services.api.projects.UnzippedProject> does not have modifier FINAL in (UnzippedProject.java:0)
Class <org.eclipse.sirius.web.services.api.representations.RepresentationDescriptor> does not have modifier FINAL in (RepresentationDescriptor.java:0)

[ERROR] immutableClassesShouldHaveANestedBuilder  Time elapsed: 0.007 s  <<< ERROR!
java.lang.IllegalArgumentException: No field with name 'projectManifest in class org.eclipse.sirius.web.services.api.projects.UnzippedProject$Builder
```

I did not modify those files.

They also seem to appear on the `main` branch. They do not appear when running `./mvnw clean verify` though, which is weird.

Also, from time to time I am getting the following compilation errors:

```
[ERROR] Tests run: 1, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 0.104 s <<< FAILURE! - in org.eclipse.sirius.web.services.validation.semantic.rules.NoDataFlowLoopsTests
[ERROR] testLoop  Time elapsed: 0.06 s  <<< ERROR!
java.lang.Error:
Unresolved compilation problems:
        The method getUnits() from the type ComputationApplicationRelease refers to the missing type EList
        The method getPins() from the type UnitCall refers to the missing type EList
        The method setSource(ConnectableDataPin) in the type DataFlow is not applicable for the arguments (Object)
        The method setTarget(ConnectableDataPin) in the type DataFlow is not applicable for the arguments (Object)
        The method getFlows() from the type ComputationApplicationRelease refers to the missing type EList

        at org.eclipse.sirius.web.services.validation.semantic.rules.NoDataFlowLoopsTests.testLoop(NoDataFlowLoopsTests.java:16)
```

They also appear in my IDE

![image](https://user-images.githubusercontent.com/889383/142254534-ab0e5a7d-b2c4-4675-81db-82b0ebbc8d64.png)

but the application and the tests work fine most of the time.

--- a/backend/sirius-web-services/src/test/java/org/eclipse/sirius/web/services/validation/semantic/rules/NoDataFlowLoopsTests.java
+++ b/backend/sirius-web-services/src/test/java/org/eclipse/sirius/web/services/validation/semantic/rules/NoDataFlowLoopsTests.java
@@ -1,0 +1,94 @@
+package org.eclipse.sirius.web.services.validation.semantic.rules;
+
+import org.junit.jupiter.api.Test;
+import static org.assertj.core.api.Assertions.assertThat;
+
+import eu.balticlsc.model.CAL.CALFactory;
+import eu.balticlsc.model.CAL.ComputationUnitRelease;
+import eu.balticlsc.model.CAL.DataBinding;
+
+public class NoDataFlowLoopsTests {
+    @Test
+    public void testLoop() {
+        var factory = CALFactory.eINSTANCE;
+        var application = factory.createComputationApplicationRelease();
+        var unitRelease = this.getTestCaseUnitRelease();
+        application.getUnits().add(unitRelease);
+
+        var unitCall = factory.createUnitCall();
+        unitCall.setName("Example Call"); //$NON-NLS-1$
+        unitCall.setUnit(unitRelease);
+
+        var unitCallPins = unitCall.getPins();
+
+        assertThat(unitCallPins).hasSize(2);
+        var inputPin = unitCallPins.stream().filter(pin -> pin.getDeclared().getBinding() == DataBinding.REQUIRED).findFirst().get();
+        var outputPin = unitCallPins.stream().filter(pin -> pin.getDeclared().getBinding() == DataBinding.PROVIDED).findFirst().get();
+
+        var dataFlow = factory.createDataFlow();
+        dataFlow.setSource(outputPin);
+        dataFlow.setTarget(inputPin);
+        application.getFlows().add(dataFlow);
+
+        var rule = new NoDataFlowLoops();
+        var diagnostics = rule.validate(application);
+
+        assertThat(diagnostics).hasSize(1);
+        var diagnostic = diagnostics.get(0);
+        assertThat(diagnostic.getMessage()).contains("Data cannot flow between pins on the same Unit Call 'Example Call'"); //$NON-NLS-1$
+    }
+
+    @Test
+    public void testNoLoop() {
+        var factory = CALFactory.eINSTANCE;
+        var application = factory.createComputationApplicationRelease();
+        var unitRelease = this.getTestCaseUnitRelease();
+        application.getUnits().add(unitRelease);
+
+        var unitCall1 = factory.createUnitCall();
+        unitCall1.setName("Example Call"); //$NON-NLS-1$
+        unitCall1.setUnit(unitRelease);
+
+        var unitCall1Pins = unitCall1.getPins();
+        assertThat(unitCall1Pins).hasSize(2);
+        var outputPin = unitCall1Pins.stream().filter(pin -> pin.getDeclared().getBinding() == DataBinding.PROVIDED).findFirst().get();
+
+        var unitCall2 = factory.createUnitCall();
+        unitCall2.setName("Example Call 2"); //$NON-NLS-1$
+        unitCall2.setUnit(unitRelease);
+
+        var unitCall2Pins = unitCall2.getPins();
+        assertThat(unitCall2Pins).hasSize(2);
+        var inputPin = unitCall2Pins.stream().filter(pin -> pin.getDeclared().getBinding() == DataBinding.REQUIRED).findFirst().get();
+
+        var dataFlow = factory.createDataFlow();
+        dataFlow.setSource(outputPin);
+        dataFlow.setTarget(inputPin);
+        application.getFlows().add(dataFlow);
+
+        var rule = new NoDataFlowLoops();
+        var diagnostics = rule.validate(application);
+
+        assertThat(diagnostics).isEmpty();
+    }
+
+    private ComputationUnitRelease getTestCaseUnitRelease() {
+        var unitRelease = CALFactory.eINSTANCE.createComputationUnitRelease();
+        unitRelease.setName("Example Unit"); //$NON-NLS-1$
+        unitRelease.setVersion("v0.1"); //$NON-NLS-1$
+
+        var inputPin = CALFactory.eINSTANCE.createDeclaredDataPin();
+        inputPin.setName("Input"); //$NON-NLS-1$
+        inputPin.setBinding(DataBinding.REQUIRED);
+
+        var outputPin = CALFactory.eINSTANCE.createDeclaredDataPin();
+        outputPin.setName("Output"); //$NON-NLS-1$
+        outputPin.setBinding(DataBinding.PROVIDED);
+
+        var pins = unitRelease.getDeclaredPins();
+        pins.add(inputPin);
+        pins.add(outputPin);
+
+        return unitRelease;
+    }
+}

